tests: migrate unit tests to Rust

dongsupark opened the pull request:
To avoid issues like python2 not available on distros, we should simply migrate unit tests to Rust. That would make more sense, as the main code is already written in Rust.

Add edition in Cargo.toml to fix builds.
Without edition = 2021, build fails due to a missing "extern crate ...".
Specifying an edition, Build passes without having to add it.

As Rust does not support global variables by default, it is necessary to make use of lazy_static, so the variables could be accessed in multiple tests.

Fixes: https://github.com/flatcar/Flatcar/issues/1081

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,499 @@
+
+#[cfg(test)]
+mod tests {
+    extern crate subprocess;
```
Extern crate shouldn't be needed, or? Normally use is enough: https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html

dongsupark replied:
I had thought also that extern crate would not be needed.
But I had to add it, due to a strange build error that suggested adding that.

Now I figured out why.
This repo is soooo ancient, ~7 years old, that its `Cargo.toml` does not have an `edition`.
Adding `edition = "2021"` fixed such a strange issue.
See also https://github.com/rust-lang/vscode-rust/issues/513#issuecomment-472393831.

pothos replied:
Right, opting into a newer edition makes sense among other renovation.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,499 @@
+            None => panic!("unable to get a current username"),
+        };
+        Popen::create(
+            [&[SCRIPT_PATH, "-u", username.into_string().unwrap().as_str(), "--ssh-dir", SSH_DIR], args.as_slice()].concat().as_slice(),
```
Instead of calling the compiled binary which will only work once `cargo build` ran, you could rather call a high-level function here that does the work of the main function. `cargo test` is expected to compile what's needed, and making it reuse some old build artifacts results in unpredictable behavior.

dongsupark replied:
Reimplemented based on high-level functions.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+            assert!(authkeystext.contains(&*TEST_KEYS[key]));
+        }
+
+        // NOTE: the tests below causes to all the tests to fail, as it asserts
+        // that all other non-contained keys not included in authorized_keys.
+        // Apparently it fails because authorized_keys still has the other keys.
```
I think there is a misunderstanding: The tests are not supposed to run after each other one by one and carry over a global state in a directory. The tests are supposed to be isolated, all having their own temporary state directory.
Quote from the Python unittest docs:

> The [setUp()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp) and [tearDown()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown) methods allow you to define instructions that will be executed before and after each test method

So, please remove the explicit number prefixes from the tests, generate a unique temporary directory in each test function (you can call the setup helper function and let it return the generated path), and remove the `test-threads=1` restriction.

dongsupark replied:
removed --test-threads=1, now it runs in parallel.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+        let aks = open_authorized_keys();
+
+        aks.write().expect("failed to update authorized keys directory");
+
```
This should call `ask.sync()` and check for the `NoKeysFound` which is expected to be the returned error as done in the python test (through stderr but we don't need this detour here):
```
 106   │         self.assertIn('no keys found', err)
```

dongsupark replied:
This is also done in a slightly different way.
It was not possible to simply compare the return value to ErrorKind, due to a trait bound issue.
So simply compare string to string.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+
+        let keys = if is_stdin {
+            // read the keys from stdin
+            AuthorizedKeys::read_keys(std::io::stdin()).unwrap_or_default()
```
I think for a test it doesn't make sense to try to read from stdin, maybe disallow this with a panic or remove the function parameter

pothos replied:
I think you can also remove the stdin test completely.

dongsupark replied:
removed the stdin test.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+    #[test]
+    fn test006_replace_one() {
+        // "update-ssh-keys --add one valid1.pub"
+        test004_add_one_file();
```
Since each test needs its own setup and teardown, I suggest not to call it here but copy the contents over

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+
+        // "update-ssh-keys --add one valid1.pub"
+        test004_add_one_file();
+
```
The below should check that `KeysExist` is returned as error kind

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+        test004_add_one_file();
+
+        // "update-ssh-keys --delete one valid1.pub"
+        del_one_ssh_key("valid1", "one");
```
The Python tests checks for `NoKeysFound` here as returned error for `.sync()`

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+            false,               // is_stdin
+            true,                // is_expected_success
+            ["valid1"].to_vec(), // assert_keys
+        );
```
I would also add a check that nothings of the bad key got stored to the folder.

dongsupark replied:
done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    #[test]
+    #[ignore] // ignore the test, as test_disable will be anyway called by test_enable() below.
```
Ignoring the test doesn't look like the right solution, I suggest you try to get it working after fixing the test setup.

dongsupark replied:
no ignore any more

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        assert!(authkeystext.starts_with("# auto-generated"));
+
+        for key in &keys {
+            if assert_no_key {
```
Isn't this switch or the naming switched?

dongsupark replied:
renamed to `assert_key_included`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        // that all other non-contained keys not included in authorized_keys.
+        // Apparently it fails because authorized_keys still has the other keys.
+
+        //        for (key, _) in &*TEST_KEYS {
+        //            if keys.contains(&key) {
+        //                continue;
+        //            }
+        //            assert!(!authkeystext.contains(&*TEST_KEYS[key]));
+        //        }
```
Is this still a problem and if yes, would it help to guard it with assert_no_key?

dongsupark replied:
Yes, it is still a problem, does not run as expected, guarding with the flag does not help.

pothos replied:
It fails because `test_add_bad` passes the wrong parameter.

dongsupark replied:
Thanks, done.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let unameosstr = users::get_current_username().unwrap_or_default();
+        let unamestr = unameosstr.to_str().unwrap_or_default();
+        let user =
+            users::get_user_by_name(&unamestr).ok_or_else(|| format!("failed to find user with name '{}'", unamestr)).unwrap_or(users::User::new(UID_CORE as uid_t, USERNAME_CORE, GID_CORE as gid_t));
```
Does it work to use the core user and if yes, should we always use it? Or should we rather fail if the user can't be resolved?

dongsupark replied:
`core` would work only in Flatcar production, not in the Ubuntu image of the CI.
Simply removed core user, and let it resolve at runtime.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
```
In `main.rs` I see that the writing/syncing is only done in for the `KeysDisabled` and `KeysExist` error, otherwise not.

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                aks.write().expect("failed to update authorized keys directory");
+                aks.sync().expect("failed to update authorized keys");
```
This should be done only once at the end after the for loop

dongsupark replied:
That is only true for sync, not for write.
Without running write before checking for key files, the whole test does not run at all.
Moved only sync out of the loop.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --add one valid1.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --add one valid1.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        );
+
+        // "update-ssh-keys --add one valid2.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        del_one_ssh_key(ctx.ssh_dir.to_str().unwrap(), "valid2", "one");
+
+        // "update-ssh-keys --add one valid1.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        }
+
+        // "update-ssh-keys --no-replace --add-force one valid2.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        test_add_one_file();
+
+        // "update-ssh-keys --add two valid2.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        test_disable();
+
+        // "update-ssh-keys --add-force two valid2.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --add one valid1.pub"
+        _ = add_one_ssh_key(
```
This supresses the error, please handle it

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+            let _ = pubfile.write_all(text.as_bytes());
+        }
+
+        test_no_keys(ssh_dir_path.to_str().unwrap());
+        test_first_run(ssh_dir_path.to_str().unwrap());
+
```
Please remove this here, the execution of this gives a different meaning to all tests using this function because the state is already modified then. Both tests should use their own `#[test]` function.

dongsupark replied:
I already went down to that route, but the whole test did not work, like:

```
thread 'tests::test_no_replace' panicked at 'failed to update authorized keys: Error(NoKeysFound("\"/tmp/test_update_ssh_keys.UrqDaXpWT31V/authorized_keys.d\""), State { next_error: None, backtrace: InternalBacktrace })'
, tests/test_update_ssh_keys.rs:235:20 
```

Tests passed only when a single test runs, or all tests run in only 1 thread.

pothos replied:
Then something else is not right, but invoking other tests before each tests and they leave their state is not the solution: The tests should keep their original semantics.

pothos replied:
I found the problem: `add_one_ssh_key` did not sync and write itself but relied on older state.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --add one valid1.pub"
+        test_add_one_file();
```
That's wrong because `test_add_one_file` operates on its own private tmpdir, you should pass `ctx`.

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --add one valid1.pub"
+        test_add_one_file();
```
That's wrong because `test_add_one_file` operates on its own private tmpdir, you should pass `ctx`.

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+
+        // "update-ssh-keys --add one valid1.pub"
+        // "update-ssh-keys --add two valid2.pub"
+        test_add_two();
```
That's wrong because `test_add_two` operates on its own private tmpdir, you should pass `ctx`.

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+
+        // "update-ssh-keys --add one valid1.pub"
+        // "update-ssh-keys --add two valid2.pub"
+        test_add_two();
```
That's wrong because `test_add_two` operates on its own private tmpdir, you should pass `ctx`.

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,524 @@
+        let ctx = setup_tests();
+
+        // "update-ssh-keys --disable two"
+        test_disable();
```
That's wrong because `test_disable` operates on its own private tmpdir, you should pass `ctx`.

pothos replied:
```suggestion
        test_disable_func(ctx);
```
Since it's also supposed to be an own test, you would have to add a wrapper for it
```
#[test]
fn test_disable() {
        let ctx = setup_tests();
        test_disable_func(ctx);
}
```

pothos replied:
Also, follow the same pattern for ` test_add_one_file` and `test_add_two`

pothos reviewed (changes requested):
It gets better but still doesn't do what the original tests do, I suggest to try to stay closer to them and keep the test logic equally simple.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,622 @@
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                // must write before checking if the file exists, because before writing,
+                // authorized_keys.d is not in sync.
+                aks.write().expect("failed to update authorized keys directory");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(!authkeyone.exists());
+
+                let mut svec: Vec<&str> = Vec::new();
+                svec.push(pubkeyname);
+                assert_has_keys(ssh_dir, svec, true);
```
I think part of the problem is that this function does too much compared to the same section in `main.rs`:
```
 118   │         Command::Delete { name } => {
 119   │             println!("Removing {}:", name);
 120   │             for key in aks.remove_keys(&name) {
 121   │                 if let AuthorizedKeyEntry::Valid { ref key } = key {
 122   │                     println!("{}", key.to_fingerprint_string());
 123   │                 }
 124   │             }
 125   │         }
 […]
  149   │     aks.write()
 150   │         .chain_err(|| "failed to update authorized keys directory")?;
 151   │     aks.sync()
 152   │         .chain_err(|| "failed to update authorized keys")?;
```
I think it's better to stick to this and move the assertions out here and into the test functions, and not as part of this helper function.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,622 @@
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
```
Like the delete helper, the `add_one_ssh_key` helper does more than what it should do but at the same time fails to call `sync` and `write`. Here what is done in `main.rs`:
```
  77   │         Command::Add {
  78   │             name,
  79   │             force,
  80   │             replace,
  81   │             stdin,
  82   │             keyfiles,
  83   │         } => {
  84   │             let keys = if stdin {
  85   │                 // read the keys from stdin
  86   │                 AuthorizedKeys::read_keys(std::io::stdin())?
  87   │             } else {
  88   │                 // keys are in provided files
  89   │                 let mut keys = vec![];
  90   │                 for keyfile in keyfiles {
  91   │                     let file = File::open(&keyfile)
  92   │                         .chain_err(|| format!("failed to open keyfile '{:?}'", keyfile))?;
  93   │                     keys.append(&mut AuthorizedKeys::read_keys(file)?);
  94   │                 }
  95   │                 keys
  96   │             };
  97   │             let res = aks.add_keys(&name, keys, replace, force);
  98   │             match res {
  99   │                 Ok(keys) => {
 100   │                     println!("Adding/updating {}:", name);
 101   │                     for key in &keys {
 102   │                         if let AuthorizedKeyEntry::Valid { ref key } = *key {
 103   │                             println!("{}", key.to_fingerprint_string());
 104   │                         }
 105   │                     }
 106   │                 }
 107   │                 Err(Error(ErrorKind::KeysDisabled(name), _)) => {
 108   │                     println!("Skipping add {} for {}, disabled.", name, config.user)
 109   │                 }
 110   │                 Err(Error(ErrorKind::KeysExist(_), _)) => {
 111   │                     println!("Skipping add {} for {}, already exists.", name, config.user)
 112   │                 }
 113   │                 _ => {
 114   │                     res.chain_err(|| "failed to add keys")?;
 115   │                 }
 116   │             }
 117   │         }
[…]
 149   │     aks.write()
 150   │         .chain_err(|| "failed to update authorized keys directory")?;
 151   │     aks.sync()
 152   │         .chain_err(|| "failed to update authorized keys")?;
```
So, `add_key_check_results` should not be called here and below. Please also add `sync` and `write`. I think that's why the tests are failing and we should stay closer to what the original tests do and what `main.rs` does.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,622 @@
+    }
+
+    fn test_first_run(ssh_dir: &str) {
+        // It is necessary to clean up before running the first run test,
+        // because old_authorized_keys will be only created when authorized_keys.d
+        // does not exist.
+        _ = fs::remove_dir_all(format!("{}/authorized_keys.d", ssh_dir));
```
```suggestion
```
With the suggested fixes there shouldn't be any old state and the test can do the same as the python test does.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,622 @@
+        assert!(!authkeys.exists());
+    }
+
+    fn test_first_run(ssh_dir: &str) {
```
```suggestion
    #[test]
    fn test_first_run() {
        let ctx = setup_tests();
```

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,622 @@
+        }
+    }
+
+    fn test_no_keys(ssh_dir: &str) {
```
```suggestion
    #[test]
    fn test_no_keys() {
    let ctx = setup_tests();
```
This should be an own test

pothos reviewed (changes requested):
Small summary: The helper functions should only do what's done in `main.rs` and not any additional checks. The sync and write also has to be added for the add helper.
The test setup should not call test functions but leave the clean state.
The calling of other tests means that we need wrapper functions for them because they should be their own tests, but then these could share the `TEST_func` main test logic.

dongsupark commented:
All addressed, but it still has a few issue, like test_enable failing due to collisions during clean up.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
+        let pubkeyvalid1 = "valid1";
+        let testnameone = "one";
+        let assert_keys = ["valid1"].to_vec();
+
+        // "update-ssh-keys --add one valid1.pub"
+        match add_one_ssh_key(
+            ssh_dir,      // ssh_dir
+            pubkeyvalid1, // pubkeyname
+            testnameone,  // testname
+            false,        // is_force
+            true,         // is_replace
+        ) {
+            Ok(keys) => add_key_check_results(ssh_dir, keys, pubkeyvalid1, testnameone, assert_keys),
+            Err(err) => panic!("update_ssh_keys --add failed {}", err),
+        }
```
This is slightly different from what the Python test does because it doesn't call `assert_has_keys` for `valid1` here.
I suggest to call `run_test_add_one_file(ctx)` here to keep the same behavior and reuse code.

pothos replied:
Got it now, the `assert_has_keys` is also run. Still, `run_test_add_one_file(ctx)` would be good instead of copying the code.

dongsupark replied:
called `run_test_add_one_file()`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+            true,         // is_force
+            true,         // is_replace
+        ) {
+            Ok(keys) => add_key_check_results(ssh_dir, keys, pubkeyvalid2, testnametwo, assert_keys),
```
Instead of `add_key_check_results` what I expect here is `assert_has_keys` as final check of the test: Python has `self.assertHasKeys('valid1', 'valid2')`.

pothos replied:
Saw now that it gets called indirectly

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+        let assert_keys = ["valid1", "valid2"].to_vec();
+
+        // "update-ssh-keys --disable two"
+        run_test_disable(ctx.clone());
```
```suggestion
        run_test_disable(&ctx);
```
Pass by reference, the cloning causes the removal race condition that fails the test

dongsupark replied:
Changed the type of the function parameter to `&PathBuf`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+
+    fn run_test_disable(ctx: TestContext) {
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
+
```
I rather would like to see a call to `run_test_add_two(&ctx)` here than the duplication of the logic.

dongsupark replied:
Called `run_test_add_two(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+    fn test_del_two() {
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
```
Same here, `run_test_add_two(&ctx)` instead of the duplication

dongsupark replied:
Called `run_test_add_two(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+        let pubkeyvalid1 = "valid1";
+        let testnameone = "one";
+        let assert_keys = ["valid1"].to_vec();
+
```
Can be replaced by a call to `run_test_add_one_file(&ctx)`

dongsupark replied:
Called `run_test_add_one_file(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+    fn test_add_two() {
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
```
Code below can be replaced by a call to `run_test_add_one_file(&ctx)`

dongsupark replied:
Called `run_test_add_one_file(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+    fn test_no_replace() {
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
```
Code below can be replaced by a call to `run_test_add_one_file(&ctx)`

dongsupark replied:
Called `run_test_add_one_file(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+            Ok(keys) => add_key_check_results(ssh_dir, keys, pubkeyvalid2, testnameone, assert_keys),
+            Err(err) => panic!("update_ssh_keys --no-replace --add-force failed {}", err),
+        }
+    }
```
The test case misses the original assertions in Python:
```
 164   │         self.assertHasKeys('valid1')
[…]
 171   │         self.assertHasKeys('valid1')
```

I.e.: `assert_has_keys(ssh_dir, ["valid1"].to_vec(), true);` and `assert_has_keys(ssh_dir, ["valid1"].to_vec(), true);`

The `add_key_check_results` function does something else. (It's not part of the original test but if you want you can keep it.)

pothos replied:
Now I see that assert_has_keys gets called by add_key_check_results, all good

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+    fn test_replace_one() {
+        let ctx = setup_tests();
+
+        let ssh_dir = ctx.ssh_dir.to_str().unwrap();
```
Code below can be replaced by a call to `run_test_add_one_file(&ctx)`

dongsupark replied:
Called `run_test_add_one_file(&ctx.ssh_dir)`.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,705 @@
+            Ok(keys) => add_key_check_results(ssh_dir, keys, pubkeyvalid1, testnameone, assert_keys),
+            Err(err) => panic!("update_ssh_keys --add failed {}", err),
+        }
+    }
```
The test case misses the original the final assertion from Python:
```
 134   │         self.assertHasKeys('valid1')
```

I.e.: `assert_has_keys(ssh_dir, ["valid1"].to_vec(), true);`

pothos replied:
Ah, now I see `add_key_check_results` chains a call to `assert_has_keys`

dongsupark commented:
Replaced tempdir with tempfile, as the tempdir crate was deprecated.

dongsupark commented:
Replaced users with uzers, as the users crate was deprecated.

dongsupark committed:
README: update CI status URL, badge, test program

Update URLs to CI status, badge, update programming language of
test programs.

dongsupark committed:
.travis.yml: delete travis config

Travis CI is not used any more. Delete.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,706 @@
+        testname: &str,
+    ) -> Result<Vec<AuthorizedKeyEntry>, ErrorKind> {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile)
+                .expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
```
This has no purpose

dongsupark replied:
Removed, thanks

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,706 @@
+        // but for some reason it simply hangs forever. For now do not run the
+        // following command to proceed to the next tests.
+
+        //        let assert_keys = ["valid1"].to_vec();
```
We need to find out why this fails. If we disable tests at our convenience there is no big benefit in having them at all.

dongsupark replied:
Resolved the issue by passing `AuthorizedKeys` via `TestContext`.

pothos replied:
Ah, the tests are different now because the opening reads the folder contents to memory already. It's ok to reopen explicitly but I would still keep the context valid, added suggestions for how to do that.

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,706 @@
+        let ssh_dir = ssh_dir_path.to_str().unwrap();
+        let pubkeybad = "bad";
+        let testnamebad = "bad";
+        let assert_keys = ["bad"].to_vec();
```
```suggestion
        let assert_keys = ["valid1"].to_vec();
```
The Python test has `self.assertHasKeys('valid1')`

dongsupark replied:
Done

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,709 @@
+        }
+
+        // Since aks has been already opened by setup_tests(), we have to close and reopen
+        // to avoid making it hang forever.
```
```suggestion
        // to re-read the contents (opening twice would make it hang forever).
```

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,709 @@
+        // Since aks has been already opened by setup_tests(), we have to close and reopen
+        // to avoid making it hang forever.
+        let _ = ctx.aks.close();
+        let aks = open_authorized_keys(ctx.ssh_dir.to_str().unwrap());
```
```suggestion
        ctx.aks = open_authorized_keys(ctx.ssh_dir.to_str().unwrap());
```

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,709 @@
+
+    #[test]
+    fn test_first_run() {
+        let ctx = setup_tests();
```
```suggestion
        let mut ctx = setup_tests();
```

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,709 @@
+        let aks = open_authorized_keys(ctx.ssh_dir.to_str().unwrap());
+
+        // equivalent of running "update-ssh-keys" without args
+        aks.write()
```
```suggestion
        ctx.aks.write()
```

pothos commented on tests/test_update_ssh_keys.rs:
```diff
@@ -0,0 +1,709 @@
+        // equivalent of running "update-ssh-keys" without args
+        aks.write()
+            .expect("failed to update authorized keys directory");
+        aks.sync().expect("failed to update authorized keys");
```
```suggestion
        ctx.aks.sync().expect("failed to update authorized keys");
```

dongsupark committed:
tests: migrate unit tests to Rust

To avoid issues like python2 not available on distros, we should simply
migrate unit tests to Rust. That would make more sense, as the main code
is already written in Rust.

Adding edition in Cargo.toml.
Without edition = 2021, build fails due to a missing "extern crate ...".
Specifying an edition, Build passes without having to add it.

As Rust does not support global variables by default, it is necessary to
make use of lazy_static, so the variables could be accessed in multiple
tests.

dongsupark commented:
Thanks.
